crates/astria-core/src/primitive/v1/mod.rs

```diff
@@ -488,6 +490,104 @@ where
     tree
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(
+    feature = "serde",
+    serde(try_from = "raw::TransactionId", into = "raw::TransactionId")
+)]
+pub struct TransactionId {
+    hash: String,
+}
+
+impl TransactionId {
+    /// Returns the hex-encoded 32-byte transaction hash as a string.
+    #[must_use]
+    pub fn get(&self) -> String {
+        self.hash.clone()
+    }
+
+    #[must_use]
+    pub fn to_raw(&self) -> raw::TransactionId {
+        raw::TransactionId {
+            hash: self.hash.clone(),
+        }
+    }
+
+    #[must_use]
+    pub fn into_raw(self) -> raw::TransactionId {
+        raw::TransactionId {
+            hash: self.hash.clone(),
+        }
+    }
+
+    /// Returns a copy of the transaction hash as a 32-byte array.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the transaction ID is not 32 bytes long. This should never happen, as the
+    /// transaction ID can only be created from a 32-byte array.
+    #[must_use]
+    pub fn as_bytes(&self) -> [u8; TRANSACTION_ID_LENGTH] {
+        hex::decode(self.hash.clone()).unwrap().try_into().unwrap()
+    }
+}
+
+impl From<TransactionId> for raw::TransactionId {
+    fn from(val: TransactionId) -> Self {
+        val.into_raw()
+    }
+}
+
+impl TryFrom<raw::TransactionId> for TransactionId {
+    type Error = TransactionIdError;
+
+    fn try_from(raw: raw::TransactionId) -> Result<Self, Self::Error> {
+        Self::try_from(raw.hash)
+    }
+}
+
+impl TryFrom<String> for TransactionId {
+    type Error = TransactionIdError;
+
+    fn try_from(hash: String) -> Result<Self, Self::Error> {
+        let inner =
+            hex::decode(hash.clone()).map_err(|err| TransactionIdError::HexDecodeError {
+                source: err,
+            })?;
+        if inner.len() != TRANSACTION_ID_LENGTH {
+            return Err(TransactionIdError::IncorrectTransactionIdLength {
+                received: inner.len(),
+            });
+        }
+        Ok(Self {
+            hash: hex::encode(inner),
+        })
+    }
+}
+
+impl From<[u8; TRANSACTION_ID_LENGTH]> for TransactionId {
+    fn from(inner: [u8; TRANSACTION_ID_LENGTH]) -> Self {
+        Self {
+            hash: hex::encode(inner),
+        }
+    }
+}
+
+impl std::fmt::Display for TransactionId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.hash)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum TransactionIdError {
```

Comment:
Please follow the rest of the codebase and keep the error variants private. Define this:

```rust
#[derive(Debug, thiserror::Error)]
#[error(transparent)]
pub struct TransactionIdErrror(TransactionIdErrorKind);

#[derive(Debug, thiserror::Error)]
enum TransactionIdErrorKind {
}
```